### PR TITLE
Alias 0.3.0

### PR DESCRIPTION
**Pull request by chewong**
---
Alias command spec: https://github.com/chewong/azure-cli-alias-extension/blob/dev/doc/az-alias-command-spec.md

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `./scripts/ci/test_static.sh` locally? (`pip install pylint flake8` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

### CODE REVIEW

File: src/alias/azext_alias/alias.py [hunk 1]
```diff
@@ -131,25 +131,29 @@ def transform(self, args):
...
             # Directly append invalid alias or collided alias
-            if not alias or alias[0] == '-' or (alias in self.collided_alias and
-                                                alias_index in self.collided_alias[alias]):
+            if not alias or alias[0] == '-' or ignore_next_iter or (alias in self.collided_alias and
+                                                                    alias_index in self.collided_alias[alias]):
                 transformed_commands.append(alias)
+                ignore_next_iter = alias and alias[0] == '-'
                 continue
+            else:
+                ignore_next_iter = False
```
Comment (troydai) [inline src/alias/azext_alias/alias.py]: Could you elaborate the logic here in the comments?

File: src/alias/azext_alias/__init__.py [hunk 2]
```diff
@@ -44,6 +72,12 @@ def alias_event_handler(_, **kwargs):
         # [:] will keep the reference of the original args
         args[:] = alias_manager.transform(args)
 
+        # Cache the reserved commands for validation later
```
[the remark this answers is not in the document]
Comment (chewong) [reply src/alias/azext_alias/__init__.py]: I don't have access to `load_cmd_tbl_func` in `custom.py` (need the entire command table for alias and command validation when the user invokes `alias create`). This is a mechanism to save the entire command table globally so `alias create` can have access to it.

File: src/alias/azext_alias/alias.py (hunk 1 above)
Comment (chewong) [reply src/alias/azext_alias/alias.py]: The logic here is to not transform named arguments. For example, if you have an alias `ls` pointing to `list`, running `az vm create -g xxx -n ls` should not yield `az vm create -g xxx -n list`,
 I think I can make it less ugly and more readable.

File: src/alias/azext_alias/__init__.py [hunk 3]
```diff
@@ -44,6 +72,12 @@ def alias_event_handler(_, **kwargs):
...
+        if args[:2] == ['alias', 'create']:
+            load_cmd_tbl_func = kwargs.get('load_cmd_tbl_func', lambda _: {})
+            global cached_reserved_commands  # pylint: disable=global-statement
```
[the remark this answers is not in the document]
Comment (chewong) [reply src/alias/azext_alias/__init__.py]: Added a class called `AliasCache` to store the reserved commands.

**Comment by troydai** [comment]
@chewong feel free to merge when @derekbekoe ack this.

File: src/alias/azext_alias/custom.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    """
+    Create an alias.
+
+    Args:
+        alias_name: The name of the alias.
+        alias_command: The command that the alias points to.
+    """
+    alias_name, alias_command = alias_name.strip(), alias_command.strip()
+    _validate_alias_name(alias_name)
+    _validate_alias_command(alias_command)
```
Comment (derekbekoe) [inline src/alias/azext_alias/custom.py]: This validation should happen in a _validators.py file.
See others in the CLI for examples.

**Comment by derekbekoe** [comment]
Please modify the description in setup.py to conform to https://github.com/Azure/azure-cli-extensions/blob/master/docs/extension_summary_guidelines.md.
Something like "Support for command aliases." would be okay.

File: src/alias/azext_alias/__init__.py [hunk 5]
```diff
@@ -8,32 +8,68 @@
...
+)
 from azext_alias._const import DEBUG_MSG_WITH_TIMING
 from azext_alias import telemetry
+from azext_alias import _help  # pylint: disable=unused-import
 
 logger = get_logger(__name__)
 
 
-class AliasExtensionLoader(AzCommandsLoader):
+class AliasCommandLoader(AzCommandsLoader):
```
[the remark this answers is not in the document]
Comment (chewong) [reply src/alias/azext_alias/__init__.py]: Changed it to `AliasExtCommandLoader`.

File: src/alias/azext_alias/custom.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    Get the current alias table.
+    """
+    try:
+        alias_table = get_config_parser()
+        alias_table.read(GLOBAL_ALIAS_PATH)
+        return alias_table
+    except Exception:  # pylint: disable=broad-except
+        # Exception is being handled by self.load_alias_table in alias.py
+        # so simply raise an empty CLIError to terminate all actions
+        raise CLIError()
```
[the remark this answers is not in the document]
Comment (chewong) [reply src/alias/azext_alias/custom.py]: Replaced this with `return get_config_parser()` so `az alias list` and other actions that depends on `get_alias_table()` will just get an empty alias table.

**Pull request merged**